### PR TITLE
[oidc] Add config to trust all certs and skip hostname verification

### LINKED ISSUE #5640: [oidc] Add Ability to Skip TLS Verify
For development purposes it would be interesting to be able to skip the tls verification of the oidc auth server.

### PR DESCRIPTION
**Pull request by fmhwong**
Add 2 new config to trust all server certificates and turn off hostname verification.

Fixes #5640

### CODE REVIEW

**Comment by sberyozkin** [comment]
@fmhwong Thanks for this PR. 

We should probably wrap these 2 properties into a `@ConfigGroup` `Tls` so that we can add more TLS related properties there going forward such as the custom key/truststore locations.

The other thing though, we will support an MTLS authentication method hopefully soon enough as discussed with @pedroigor, so, we should consider now IMHO how these 2 new properties will be used along the MTLS method, @pedroigor, what do you think ?

**Comment by sberyozkin** [comment]
@fmhwong Can you please one of the 4 OIDC related integration test suites, lets choose `integration-tests/oidc` which is simplest, and update it to have Keycloak available at `HTTPS` ? I'm assuming it would require using the new properties and as such will test they are effective :-) Give it a try please.
You can run the test locally in `integration-tests/oidc` like this:
```
mvn clean install -Pdocker-keycloak -Ptest-keycloak
```

**Comment by fmhwong** [comment]
Thanks for the comments! Will add a new `@ConfigGroup` and take a look of the test cases.

**Comment by pedroigor** [comment]
> @fmhwong Thanks for this PR.
> 
> We should probably wrap these 2 properties into a `@ConfigGroup` `Tls` so that we can add more TLS related properties there going forward such as the custom key/truststore locations.

I think it makes sense to have such group so that users can set any global TLS related config.

Btw, what if we have just a single `hostname-verification` option and use a enum to define the behavior ? For instance, `hostname-verification=NONE` would disable it, etc.

> 
> The other thing though, we will support an MTLS authentication method hopefully soon enough as discussed with @pedroigor, so, we should consider now IMHO how these 2 new properties will be used alongside the MTLS method, @pedroigor, what do you think ?

For MTLS, I think this group is going to provide some core properties such as keystore/password/key, so that web type clients can authenticate to the authorization server using certificates. For service type clients, the config group should help to establish the connection to clients so that they can ask their certificates and validate certificate-bound access tokens.

Basically, two main use cases for the different types of applications we support.

**Comment by sberyozkin** [comment]
Hi @pedroigor Thanks, this is helpful...
Sure, making `hostname-verification` an enum can be good, my understanding though it is only either on or off as far as the VertxOAuth2 is concerned. Please comment once @fmhwong introduces a group.
Thanks

**Comment by fmhwong** [comment]
@pedroigor  Just want to clarify that you are suggesting to combine the proposed 2 new config `trustAllCertificates` and `hostnameVerification` into a single config `hostnameVerification` and use `quarkus.oidc.tls.hostname-verification=NONE` to trust all certs and disable hostname verification.

**Comment by sberyozkin** [comment]
@fmhwong Hi, can you please wrap those properties in a group which is what Pedro is also fine with and force-push and then we will all discuss if/how the properties can be optimized, we will grow that `Tls` group anyway going forward. Cheers

**Comment by fmhwong** [comment]
@sberyozkin Just did.  Please take a look. Thanks!

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcTenantConfig.java
Comment (sberyozkin) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcTenantConfig.java]: @fmhwong Please change it to `Tls tls = new Tls()` and rename the group to `Tls` so that we can have `quarkus.oidc.tls.trust-all-certificates`, etc, the `config` is assumed :-)

**Comment by sberyozkin** [comment]
@pedroigor Hey Pedro, OK, the group is here, can you clarify please what you'd like to change ? Cheers

**Comment by pedroigor** [comment]
@sberyozkin @fmhwong I think that we could squash both options in `tls` into a single one. One of the reasons is that they are quite related when you want to disable TLS verifications (hostname and certs). Another reason is that we could eventually support more options on how verification should work. I'm mainly saying that due to the fact that some HTTP client libraries provide different options to set how verification is done (e.g.: apache http client). So I think it may be useful and more flexible.

**Comment by sberyozkin** [comment]
@pedroigor OK, so `Tls` group will have an enum (`Verification` ?), with `REQUIRED` and `NONE` being two initials values, something like this:
```
class Tls {
   enum Verification {
        // Default: the hostname verification is enabled
        REQUIRED,
        // the hostname verification is disabled, all certs are trusted
        NONE
        // other options as needed (trust specific hosts etc) ?       
   }
  @ConfigItem(defaultValue=Verification.REQUIRED) 
  Vericaition verification; 
}
```

something like that ?

**Comment by fmhwong** [comment]
Thanks and I will make the change.

**Comment by fmhwong** [comment]
I noticed the config already has `quarkus.vault.tls.skip-verify`.  Do you want to name this `quarkus.oidc.tls.skip-verify` and set the type to `boolean`?  Default to `false` and set to `true` to trust all certs and skip host verification.

**Comment by fmhwong** [comment]
@sberyozkin @pedroigor What do you think?

**Comment by sberyozkin** [comment]
@fmhwong We don't have it to align it with the `Vault` extension IMHO as it is the only property there and we'd like to keep the options more diverse with the `Tls` group.

**Comment by pedroigor** [comment]
> @sberyozkin @pedroigor What do you think?

@fmhwong Good catch. Align sounds better than introducing a different way of doing it.

**Comment by sberyozkin** [comment]
@pedroigor I'm sorry, I'm not following. That `Vault` property is a pretty basic way to configure it for Vault, why would we want to limit the configuration to a single boolean property now ? Would it also be a step back from what you proposed earlier (we have more flexible TLS support, etc) ?

**Comment by sberyozkin** [comment]
If we were go back from the enum idea then I'd just stick to the current PR:
```
    @ConfigGroup
    public static class TlsConfig {
        /**
         * If all server certificates should be trusted
         */
        @ConfigItem(defaultValue = "false")
        boolean trustAllCertificates;
        /**
         * If hostname verification is enabled
         */
        @ConfigItem(defaultValue = "true")
        boolean hostnameVerification;
    }
```
as these are 2 distinctive properties.

**Comment by pedroigor** [comment]
@sberyozkin Yes, it is a step back ... I did not know we already had this in Quarkus.

My proposal is mainly based on what we currently have in Keycloak in regards to this type of verification. There we have an enum. So I thought that future-wise it would be nice to protect us from future changes.

But now that we know we have the same capability in another extension, it makes more sense to me to have the same. I do prefer the enum option, but in favor of a better UX, I'm OK to review my preference.

I also think the Vault approach is better than have two options. As we discussed, most of the times you are going to disable boths, right ?

**Comment by sberyozkin** [comment]
@pedroigor I thought when you suggested to use an enum and mentioning it was more flexible that we'd want to support more than just `true`, `false` with respect to `Tls` verification.
My understanding is that right now, in Quarkus, different extensions where the HTTPS client invocations are involved control things differently. Having a uniform approach across all of the extensions would be a worthy idea to try to realize but it would require a dedicated efforts from many of us :-). In meantime even if we name this property the same as in Vault, users who use both Vault and OIDC would still have to type this property twice :-)

May be we need to back to a drawing board :-).

So this `Tls` group is a start where we will control different things related to `Tls` including `Mtls`.
We have several requirements, possibly somewhat orthogonal but all related to `Tls`:
1. Disable/enable the host/cert verification (this PR)
2. We have #7543 (I suppose we'd need to support setting the custom trustore/keystore ?)
These are to support the discovery and most of the client autentication methods over TLS
3.  and Client authentication MTLS method will need to be added in time
4. Do we want to support some specific hostname verification options ?

So `NONE/REQUIRED` enums lets us do 1. Or we can have instead `quarkus.oidc.tls.skip-verify`.
It also works.

The question is how we can move from there when addressing other requirements above.
Do you think `quarkus.oidc.tls.skip-verify` will let us build upon it ? Or enums are better ? 

`quarkus.oidc.tls.verification=none` or `quarkus.oidc.tls.verification=required` (default) (note Quarkus is fine with converting lower case values like `none` to `NONE`/etc :-) ) reads quite good to me :-)

**Comment by pedroigor** [comment]
@sberyozkin OK, I agree to the `.verification=none|required`.

**Comment by fmhwong** [comment]
Thanks guys! I will make the changes.

**Comment by sberyozkin** [comment]
@pedroigor thanks, I don't really want to push for it, if you also feel OK about it then it is good. I'll CC Guillaume as he knows a lot about these TLS issues, @gsmet, Guillaume (and others), comments are welcome

**Comment by sberyozkin** [comment]
@fmhwong Thanks, can you please update the test as agreed earlier ?

**Comment by fmhwong** [comment]
Yes, will do

**Comment by sberyozkin** [comment]
@fmhwong Hi, thanks for your effort, it looks good :-). But you can simplify. We have enough non-secure `http` already in various integration tests, so if you just update `integration-tests/oidc` then it should be fine IMHO, only update URLs to `https:` and disable the verification in the application properties and it will be enough :-)
Thanks again for spending the time on it

**Comment by sberyozkin** [comment]
@fmhwong Thanks, IMHO it is ready to go as soon as CI passes. Please also squash all your commits.
Thanks

**Comment by sberyozkin** [comment]
@fmhwong Hi, thanks for getting the tests fixed, the native build failed with the unrelated issue. You have definitely helped with starting the TLS related enhancements, it would've taken few cycles to trace that the ci actions have to be fixed, etc, nice work.
Can you please squash all your commits ?

**Comment by fmhwong** [comment]
@sberyozkin I squashed the commits and all checks have passed.  Thanks for your help!

**Comment by sberyozkin** [comment]
@fmhwong Thanks.

Now, before I merge, I'd like to circle a bit. I've been thinking during the last few days if your @fmhwong idea to align with Vault and just have `boolean skip-verify` should be revisited, something @pedroigor also thought was worth an attention. IMHO it is OK to consider it again, I certainly do it all the time with my PRs :-)

I've looked at how Vertx handles these 2 booleans.
`HttpClientOptions` deals with the host verification while `ClientOptionsBase` deals with `trustAll` - if it is false then it checks the trust options. But if we were to add the trust options then it would make `trustAll=false` anyway. I've been thinking that the enums will give us more options based on the original discussion with Pedro, but I'm not really seeing what else we can actually add to these `NONE` and `REQUIRED` now.

So I'd like to hold off the merge a bit. @fmhwong @pedroigor, can you think of at least one more useful enum value ? If yes - then we should keep the enum. If not lets just do `boolean skip-verify` :-)

Thanks all

**Review by sberyozkin — APPROVED** [review]
LGTM, but lets do one final discussion around the property names (use enum vs a single `boolean skipVerify`)

**Comment by sberyozkin** [comment]
I'm going to go ahead with the merge later today if there will be no additional comments, the enums are more descriptive so it is still a good approach, I'd just like to give a bit more time to think about it.

**Pull request merged**
